compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -490,13 +490,97 @@ static bool isIndexVariable(Symbol* sym, CForLoop* loop) {
   return false;
 }
 
-static  CallExpr* generateGPUCall(FnSymbol* kernel,
-                                     std::vector<VarSymbol*> actuals) {
+static Symbol* extractUpperBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->testBlockGet())) {
+      if(CallExpr *call = toCallExpr(bs->body.head)) {
+        if(call->isPrimitive(PRIM_LESSOREQUAL)) {
+          if(SymExpr *symExpr = toSymExpr(call->get(2))) {
+            return symExpr->symbol();
+          }
+        }
+      }
+  }
+
+  INT_FATAL("Unexpected upper bound for loop identified for extraction as GPU kernel");
+  return nullptr;
+}
+
+static Symbol* extractLowerBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->initBlockGet())) {
+    for_alist(entry, bs->body) {
+      if(CallExpr *call = toCallExpr(entry)) {
+        if(call->isPrimitive(PRIM_ASSIGN)) {
+          if(SymExpr *symExpr = toSymExpr(call->get(2))) {
+            return symExpr->symbol();
+          }
+        }
+      }
+    }
+  }
+
+  INT_FATAL("Unexpected lower bound for loop identified for extraction as GPU kernel");
+  return nullptr;
+}
+
+static VarSymbol* generateNewVariable(BlockStmt* insertionPoint, const char* name, Type* type) {
```

Comment:
Could you rename this to `insertNewVariable` or `insertNewDef` or something? I want the name to convey that it changes AST and not just returns something new.

Reply:
How about `insertNewVarAndDef`?

Reply:
Yep, that's good.